Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,358 @@
+
+## TODO
+
+* Treat the noncommutative case
+
+* Treat the case of modules over semirings
+  (For a possible definition of an exact sequence of commutative semigroups, see
+  [Grillet-1969b], Pierre-Antoine Grillet,
+  *The tensor product of commutative semigroups*,
+  Trans. Amer. Math. Soc. 138 (1969), 281-293, doi:10.1090/S0002-9947-1969-0237688-1 .)
+
+* Treat algebras (further PR)
+-/
+
+
+section Modules
+
+open TensorProduct LinearMap
+
+section Semiring
+
+variable {R : Type*} [CommSemiring R]
+variable {M N P P' Q: Type*}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommGroup P] [AddCommMonoid P'] [AddCommMonoid Q]
+  [Module R M] [Module R N] [Module R P] [Module R P'] [Module R Q]
+
+variable {f : M →ₗ[R] N} (g : N →ₗ[R] P)
+
+lemma le_comap_range_lTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (lTensor Q g)).comap (TensorProduct.mk R Q P q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨q ⊗ₜ[R] n, rfl⟩
+
+lemma le_comap_range_rTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (rTensor Q g)).comap
+      ((TensorProduct.mk R P Q).flip q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨n ⊗ₜ[R] q, rfl⟩
+
+variable (Q) {g}
+
+/-- If `g` is surjective, then `lTensor Q g` is surjective -/
+theorem lTensor.surjective (hg : Function.Surjective g) :
+    Function.Surjective (lTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | zero => exact ⟨0, map_zero _⟩
+  | tmul q p =>
+    obtain ⟨n, rfl⟩ := hg p
+    exact ⟨q ⊗ₜ[R] n, rfl⟩
+  | add x y hx hy =>
+    obtain ⟨x, rfl⟩ := hx
+    obtain ⟨y, rfl⟩ := hy
+    exact ⟨x + y, map_add _ _ _⟩
+
+/-- If `g` is surjective, then `rTensor Q g` is surjective -/
+theorem rTensor.surjective (hg : Function.Surjective g) :
+    Function.Surjective (rTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | zero => exact ⟨0, map_zero _⟩
+  | tmul p q =>
+    obtain ⟨n, rfl⟩ := hg p
+    exact ⟨n ⊗ₜ[R] q, rfl⟩
+  | add x y hx hy =>
+    obtain ⟨x, rfl⟩ := hx
+    obtain ⟨y, rfl⟩ := hy
+    exact ⟨x + y, map_add _ _ _⟩
+
+end Semiring
+
+variable {R : Type*} [CommRing R]
+variable {M N P P' : Type*}
+  [AddCommGroup M] [AddCommGroup N] [AddCommGroup P]
+  [Module R M] [Module R N] [Module R P]
+
+open Function LinearMap
+
+lemma LinearMap.exact_subtype_mkQ (Q : Submodule R N) :
+    Exact (Submodule.subtype Q) (Submodule.mkQ Q) := by
+  rw [exact_iff, Submodule.ker_mkQ, Submodule.range_subtype Q]
+
+lemma LinearMap.exact_map_mkQ_range (f : M →ₗ[R] N) :
+    Exact f (Submodule.mkQ (range f)) :=
+  exact_iff.mpr <| Submodule.ker_mkQ _
+
+lemma LinearMap.exact_subtype_ker_map (g : N →ₗ[R] P) :
+    Exact (Submodule.subtype (ker g)) g :=
+  exact_iff.mpr <| (Submodule.range_subtype _).symm
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+variable (Q : Type*) [AddCommGroup Q] [Module R Q]
+
+variable (hfg : Exact f g) (hg : Function.Surjective g)
+
+
+/-- The direct map in `lTensor.equiv` -/
+def lTensor.toFun :
```

Comment:
How should I think of this name compared to `LinearMap.lTensor.toFun`? @riccardobrasca

Reply:
There is no direct relation.
 `LinearMap.lTensor f` is the linear map `Q tensor N -> Q tensor P` associated with a linear map `g : N -> P`.
Here, we want to understand this map `LinearMap.lTensor g` when we know that `g` is surjective and know its kernel : it is surjective (by `lTensor.surjective`) and we compute its kernel in terms of a map `f : M -> N` whose image is the kernel of `g`.